### ISSUE TITLE
Generic class constructors are not properly identified as functions

### ISSUE THREAD

**Issue by alexchandel** [python/mypy#15779]
**Bug Report**

The constructors of generic classes are not properly identified as functions.

**To Reproduce**

```python
#!/usr/bin/env python3
"""Test."""
from typing import Generic, TypeVar

_T = TypeVar('_T')


class Wrapper(Generic[_T]):
    """Test."""

    def __init__(self, x: _T):
        """Test."""
        self.x = x


foo: Wrapper[str]
bar: Wrapper[str]

foo, bar = map(Wrapper, ['foo', 'bar'])
```

[playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=ce5775fa2ffdd83148832fa1277cdee7)

**Expected Behavior**

The `str` are mapped to `Wrapper[str]`.

**Actual Behavior**

`mypy` fails to work out the type parameter of `Wrapper`, and (probably related) fails to work out that `Wrapper`'s constructor is a `Callable[[str], Wrapper[str]]`.

```
main.py:19: error: Incompatible types in assignment (expression has type "Wrapper[_T]", variable has type "Wrapper[str]")  [assignment]
main.py:19: error: Argument 1 to "map" has incompatible type "type[Wrapper[Any]]"; expected "Callable[[str], Wrapper[_T]]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

Latest mypy 1.4.1.  Latest python 3.11.4.

**Comment by ilevkivskyi** [comment]
This should work with `--new-type-inference`. I am going to close this, since there is no action items for us (except enabling this flag by default, but we will do this anyway around September).

**Comment by alexchandel** [comment]
Where's the documentation for that flag?  I don't see it in `-h` anywhere.

**Comment by JelleZijlstra** [comment]
It's intentionally undocumented while the feature is being developed.

**Comment by alexchandel** [comment]
So when I use `--new-type-inference`, `mypy` seems to stop linting altogether.  Is it nonfunctional?

**Comment by ilevkivskyi** [comment]
What do you mean by "stops linting altogether"? Is there some false negative (code that crashes but not caught by mypy when that flag is on)?

**Comment by alexchandel** [comment]
As in there is no output at all.  It took me a moment to notice!

**Comment by JelleZijlstra** [comment]
That is very odd. Could it be that there is a segfault or something and your OS doesn't print anything in that case?
